### PR TITLE
drivers/counter: stm32: Convert configuration from Kconfig to device tree

### PR DESCRIPTION
**Pull request by gautierg-st**
Use device tree to configure rtc counter driver instead of Kconfig for:
- RTC source clock selection (thanks to use of dts domain clock)
- LSE drive capability (use lse dts property)

### CODE REVIEW

File: include/zephyr/dt-bindings/clock/stm32f0_clock.h [hunk 1]
```diff
@@ -27,6 +27,8 @@
 /** PLL clock */
 #define STM32_SRC_PLLCLK	0x006
 
+#define STM32_SRC_LSI		0x007
```
Comment (erwango) [inline include/zephyr/dt-bindings/clock/stm32f0_clock.h:30]: Please move to `/** Fixed clocks  */` with LSE/HSI.
Apply everywhere applicable.

File: drivers/counter/counter_ll_stm32_rtc.c [hunk 2]
```diff
@@ -406,55 +430,55 @@ static int rtc_stm32_init(const struct device *dev)
...
-	* && !CONFIG_SOC_SERIES_STM32L1X
-	*/
+		!defined(CONFIG_SOC_SERIES_STM32F2X) && \
+		!defined(CONFIG_SOC_SERIES_STM32F1X) && \
+		!defined(CONFIG_SOC_SERIES_STM32L1X)
+
+		/* Configure driving capability */
+		LL_RCC_LSE_SetDriveCapability(STM32_LSE_DRIVING << RCC_BDCR_LSEDRV_Pos);
+
+#endif		/*
```
Comment (erwango) [inline drivers/counter/counter_ll_stm32_rtc.c:455]: Use one space instead of tab here.

File: drivers/counter/counter_ll_stm32_rtc.c [hunk 3]
```diff
@@ -406,55 +430,55 @@ static int rtc_stm32_init(const struct device *dev)
...
+		LL_RCC_LSI1_Enable();
+		while (LL_RCC_LSI1_IsReady() != 1) {
+		}
 #else
-	LL_RCC_LSI_Enable();
-	while (LL_RCC_LSI_IsReady() != 1) {
-	}
+		LL_RCC_LSI_Enable();
+		while (LL_RCC_LSI_IsReady() != 1) {
+		}
```
Comment (erwango) [inline drivers/counter/counter_ll_stm32_rtc.c:441]: Clocks activation should now be performed y clock_control driver and hence removed from here.

File: drivers/counter/counter_ll_stm32_rtc.c [hunk 4]
```diff
@@ -406,55 +430,55 @@ static int rtc_stm32_init(const struct device *dev)
...
-	* && !CONFIG_SOC_SERIES_STM32F2X
-	* && !CONFIG_SOC_SERIES_STM32F1X
-	* && !CONFIG_SOC_SERIES_STM32L1X
-	*/
+		!defined(CONFIG_SOC_SERIES_STM32F2X) && \
+		!defined(CONFIG_SOC_SERIES_STM32F1X) && \
+		!defined(CONFIG_SOC_SERIES_STM32L1X)
+
+		/* Configure driving capability */
+		LL_RCC_LSE_SetDriveCapability(STM32_LSE_DRIVING << RCC_BDCR_LSEDRV_Pos);
```
Comment (erwango) [inline drivers/counter/counter_ll_stm32_rtc.c:453]: This should also be done by clock_control driver

File: drivers/counter/counter_ll_stm32_rtc.c [hunk 5]
```diff
@@ -406,55 +430,55 @@ static int rtc_stm32_init(const struct device *dev)
...
+		/* Wait until LSE is ready */
+		while (LL_RCC_LSE_IsReady() != 1) {
+		}
 
 #if STM32_MSI_PLL_MODE
-	/* Enable MSI hardware auto calibration */
-	LL_RCC_MSI_EnablePLLMode();
+		/* Enable MSI hardware auto calibration */
+		LL_RCC_MSI_EnablePLLMode();
 #endif
```
Comment (erwango) [inline drivers/counter/counter_ll_stm32_rtc.c:475]: Ditto

File: drivers/counter/counter_ll_stm32_rtc.c [hunk 6]
```diff
@@ -57,9 +57,23 @@ LOG_MODULE_REGISTER(counter_rtc_stm32, CONFIG_COUNTER_LOG_LEVEL);
...
+	{.bus = STM32_SRC_LSE, .enr = RTC_SEL(1)}
+#endif
+};
+#else
+static const struct stm32_pclken rtc_clk[] = STM32_DT_INST_CLOCKS(0);
+#endif
+
 struct rtc_stm32_config {
 	struct counter_config_info counter_info;
-	struct stm32_pclken pclken;
```
Comment (erwango) [inline drivers/counter/counter_ll_stm32_rtc.c:62]: Why removing clock_config from this structure ?

**Review by erwango — CHANGES_REQUESTED** [review]

File: include/zephyr/dt-bindings/clock/stm32f0_clock.h (hunk 1 above)
Comment (gautierg-st) [reply include/zephyr/dt-bindings/clock/stm32f0_clock.h]: Done

**Review by mbolivar-nordic — COMMENTED** [review]
ACK on the bindings changes

File: include/zephyr/dt-bindings/clock/stm32f4_clock.h [hunk 7]
```diff
@@ -26,5 +26,43 @@
 #define STM32_SRC_PLL_P	0x001
 #define STM32_SRC_PLL_Q	0x002
 #define STM32_SRC_PLL_R	0x003
+#define STM32_SRC_LSE	0x004
+#define STM32_SRC_LSI	0x005
```
Comment (erwango) [inline include/zephyr/dt-bindings/clock/stm32f4_clock.h:30]: /** Fixed clocks */

File: drivers/counter/counter_ll_stm32_rtc.c [hunk 8]
```diff
@@ -392,69 +392,19 @@ static int rtc_stm32_init(const struct device *dev)
...
-#if STM32_MSI_PLL_MODE
-	/* Enable MSI hardware auto calibration */
-	LL_RCC_MSI_EnablePLLMode();
-#endif
-
-	LL_RCC_SetRTCClockSource(LL_RCC_RTC_CLKSOURCE_LSE);
-
-#endif /* CONFIG_COUNTER_RTC_STM32_CLOCK_SRC */
+	/* Enable RTC clock source */
+	clock_control_configure(clk, (clock_control_subsys_t *) &cfg->pclken[1],
```
Comment (erwango) [inline drivers/counter/counter_ll_stm32_rtc.c:406]: Add error checking.

File: drivers/counter/counter_ll_stm32_rtc.c [hunk 9]
```diff
@@ -494,40 +444,52 @@ static int rtc_stm32_init(const struct device *dev)
...
+static const struct stm32_pclken rtc_clk[] = {
+	STM32_CLOCK_INFO(0, DT_DRV_INST(0)),
+	/* Use Kconfig to configure source clocks fields */
+	/* Fortunately, values are consistent across enabled series */
+#ifdef COUNTER_RTC_STM32_CLOCK_LSI
+	{.bus = STM32_SRC_LSI, .enr = RTC_SEL(2)}
+#else
+	{.bus = STM32_SRC_LSE, .enr = RTC_SEL(1)}
+#endif
+};
```
Comment (erwango) [inline drivers/counter/counter_ll_stm32_rtc.c:457]: Add a warning in this case which is now deprecated.

File: drivers/counter/counter_ll_stm32_rtc.c [hunk 10]
```diff
@@ -392,69 +392,19 @@ static int rtc_stm32_init(const struct device *dev)
...
 		LOG_ERR("clock op failed\n");
 		return -EIO;
 	}
 
+	/* Enable Backup access */
 	z_stm32_hsem_lock(CFG_HW_RCC_SEMID, HSEM_LOCK_DEFAULT_RETRY);
-
 	LL_PWR_EnableBkUpAccess();
 
-#if defined(CONFIG_COUNTER_RTC_STM32_BACKUP_DOMAIN_RESET)
```
Comment (erwango) [inline drivers/counter/counter_ll_stm32_rtc.c:404]: Kconfig symbol should be cleaned up.

**Review by erwango — COMMENTED** [review]
Some comments remaining otherwise looks good.
Additionally, please update release notes (3.3) with deprecated and removed Kconfig symbols

File: doc/releases/release-notes-3.3.rst [hunk 11]
```diff
@@ -189,6 +206,14 @@ USB
 Devicetree
 **********
 
+* Bindings
+
+  * Other changes:
+
+    * STM32 SoCs:
+
+      * LSE Clock binding: new ``lse-bypass`` property
```
Comment (erwango) [inline doc/releases/release-notes-3.3.rst:215]:
```suggestion
      * :dtcompatible: `st,stm32-lse-clock`: new ``lse-bypass`` property
```

**Review by erwango — DISMISSED** [review]

**Review by erwango — DISMISSED** [review]

**Review by erwango — APPROVED** [review]

File: boards/arm/nucleo_l152re/nucleo_l152re.dts [hunk 12]
```diff
@@ -85,6 +89,8 @@
 };
 
 &rtc {
+	clocks = <&rcc STM32_CLOCK_BUS_APB1 0x10000000>,
```
Comment (FRASTM) [inline boards/arm/nucleo_l152re/nucleo_l152re.dts:92]: Check RM0394 for the stm32l41x/42x/43x/44x/45x/46x  : clocks = <&rcc STM32_CLOCK_BUS_APB1 **0x400**>,

**Review by FRASTM — COMMENTED** [review]
Actually in several stm32 devices, the RTC peripheral has a Enable bit in the RCC BDCR register plus a RTC  clock  source selection on RTCSEL bits where default value 00 is "no clock" without RCC bus clock control bit.
(ex on the stm32G4 serie).

Here, the RTC node in the DTS is configured with  RTC SEL is LSI. Consequently the LSI must also be enabled.
--> That will now make the RTC running on each stm32 platform (and probably increase the power consumption).

**Review by FRASTM — APPROVED** [review]

**Pull request merged**